From 11a3e6d831d6334cba70452ea8b5eac5226686af Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Fri, 04 Oct 2024 17:34:00 +0300
Subject: [PATCH] [FSW-10557] fs1: configure gp_msb_0

With this patch added support to configure
msb of general purpose region 0, which is
required when arc try to access gen memory.

Change-Id: Ie8b179627b890706d8a711384a41c72a5a0f57cd
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 86dd2f8..340fe13 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -28,6 +28,7 @@
 #include <psoc_pad_conf_regs.h>
 #include <gpio_regs.h>
 #include <psoc_boot_conf_masks.h>
+#include <arc_regs.h>
 
 extern const int fs1_die_type;
 
@@ -463,3 +464,8 @@
 		(regmap_readl(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_BOOT_STRAP_PINS_H)) &
 			PSOC_BOOT_CONF_BOOT_STRAP_PINS_H_EFUSE_OVERRIDE_M));
 }
+
+void boot_rom_config_arc_regions(void)
+{
+	regmap_writel(FULL_ADDR_MSB, IO0_H0_MNG_ARC(mmARC_GP_MSB_ADDR_0));
+}
diff --git a/arch/fs1/arc/soc.h b/arch/fs1/arc/soc.h
index b908140..15195c2 100644
--- a/arch/fs1/arc/soc.h
+++ b/arch/fs1/arc/soc.h
@@ -94,6 +94,7 @@
 #define PCIE_FWSRAM_OFFSET		0x0380000
 #define PCIE_FWSRAM_SIZE		SZ_64K
 
+#define FULL_ADDR_MSB			0x300007F
 #define FW_MEM_BASE_FULL		0x300007FF8800000ull
 #define SCRATCHPAD_SRAM_BASE		(HL_FW_MEM_BASE + SCRATCHPAD_SRAM_OFFSET)
 #define SCRATCHPAD_SRAM_BASE_FULL	(FW_MEM_BASE_FULL + SCRATCHPAD_SRAM_OFFSET)
diff --git a/common/common.c b/common/common.c
index afabb2f..fda1ffe 100644
--- a/common/common.c
+++ b/common/common.c
@@ -55,6 +55,9 @@
 
 	boot_rom_status_update(PROGRESS_INIT);
 
+	/* Before this, we cannot have any access to gen mem region */
+	boot_rom_config_arc_regions();
+
 	boot_rom_init_ts();
 
 	boot_rom_uart_init();
diff --git a/common/common.h b/common/common.h
index 822a332..b0e7491 100644
--- a/common/common.h
+++ b/common/common.h
@@ -37,5 +37,6 @@
 void boot_rom_set_manifest_index(enum hl_fw_manifest_type manifest_idx);
 void boot_rom_start_cpld_wdtimer(void);
 bool boot_rom_spi_efuse_enabled(void);
+void boot_rom_config_arc_regions(void);
 
 #endif /* _COMMON_H_ */
